services/src/main/java/com/linkedin/kmf/services/TopicRebalancer.java

```diff
@@ -0,0 +1,427 @@
+   * Each time this is invoked zero or more of the following happens.  T
+   *
+   * 1. if number of partitions falls below threshold then create new partitions.  The produce service will need to
+   *   detect this this condition and create new metrics for the new partitions.
+   * 2. This is done during rebalance.
+   *   if a broker does not have enough partitions then assign it partitions from brokers that have excess
+   *   if a broker is not a leader of some partition then make it a leader of some partition by finding out
+   *     which broker is a leader of more than one partition.  Does this involve moving partitions among brokers?
+   * 3. run a preferred replica election
+   * </pre>
+   */
+  void rebalanceMonitoredTopic(TopicState topicState) {
+    try {
+      LOG.debug(_serviceName + ": broker count " + topicState._allBrokers.size() + " partitionCount " + topicState._partitionInfo.size() + ".");
+
+      if (topicState.partitionsLow(_expectedPartitionBrokerRatio)) {
+        int idealPartitionCount = _partitionsPerBroker * topicState._allBrokers.size();
+        int addPartitionCount = idealPartitionCount - topicState._partitionInfo.size();
+        LOG.info(_serviceName + ": adding " + addPartitionCount + " partitions.");
+        addPartitions(_zkUtils, idealPartitionCount);
+        waitForAddedPartitionsToBecomeActive(idealPartitionCount);
+        topicState = topicState();
+      }
+      if (topicState.brokerMissingPartition()) {
+        LOG.info(_serviceName + ": rebalancing monitored topic.");
+        waitForOtherAssignmentsToComplete();
+        reassignPartitions(topicState._allBrokers, topicState._partitionInfo.size());
+        waitForPartitionReassignmentToComplete();
+        topicState = topicState();
+      }
+      if (topicState.brokerNotElectedLeader()) {
+        LOG.info(_serviceName + ": running preferred replica election.");
+        runPreferredElection(_zkUtils, topicState._partitionInfo);
+      }
+    } catch (InterruptedException ie) {
+      throw new IllegalStateException(ie);
+    }
+  }
+
+  /**
+   *  Polling...
+   */
+  private void waitForAddedPartitionsToBecomeActive(int expectedNumberOfPartitions) {
+    long timeout = System.currentTimeMillis() + 1000 * 60;
+
+    //Using ZkUtils instead of the consumer because the consumer seems to cache the last answer it got which was the old
+    //number of partitions.
+    scala.collection.mutable.ArrayBuffer<String> scalaTopic = new scala.collection.mutable.ArrayBuffer<>();
+    scalaTopic.$plus$eq(_topic);
+    while (System.currentTimeMillis() < timeout) {
+      if (_zkUtils.getPartitionAssignmentForTopics(scalaTopic).apply(_topic).size()  == expectedNumberOfPartitions) {
+        return;
+      }
+      try {
+        Thread.sleep(1000);
+      } catch (InterruptedException e) {
+        throw new IllegalStateException(e);
+      }
+    }
+    throw new IllegalStateException("Waiting for additional partitions to appear timed out.");
```

Comment:
It maybe more informative to say "Partition expansion can not finish in 60 sec". Besides, I am not sure if it is appropriate to use IllegalStateException here.